subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c

```diff
@@ -0,0 +1,540 @@
+		}
+		/* Map end */
+		if (ok) {
+			ok = zcbor_map_end_encode(zse, map_count);
+		}
+
+		if (!ok) {
+			LOG_ERR("Failed to encode Image Upload packet");
+			smp_packet_free(nb);
+			image_upload_buf->status = MGMT_ERR_ENOMEM;
+			goto end;
+		}
+
+		offset_before_send = upload_state.offset;
+		nb->len = zse->payload - nb->data;
+		k_sem_reset(&spm_mutex);
+		rc = smp_client_send_cmd(nb, image_upload_res_fn, &spm_mutex, 15);
+		if (rc) {
+			LOG_ERR("Failed to send SMP Upload init packet, err: %d", rc);
+			smp_packet_free(nb);
+			image_upload_buf->status = rc;
+			goto end;
+
+		}
+		k_sem_take(&spm_mutex, K_FOREVER);
+		if (image_upload_buf->status) {
+			LOG_ERR("Upload Fail:%d", image_upload_buf->status);
+			goto end;
+		} else if (offset_before_send == upload_state.offset) {
+			/* Need to re send packet again write operation have been failed */
+		} else if (offset_before_send + write_length ==  upload_state.offset) {
+			/* Send operation normal state */
+			upload_state.writed_length += write_length;
+		} else  {
+			/* Stop this packet send offset is at further than this current packet */
+			goto end;
+		}
+	}
+end:
+	rc = image_upload_buf->status;
+	k_mutex_unlock(&mcumgr_mutex);
+	image_upload_buf = NULL;
+	return rc;
+}
+
+int img_mgmt_client_state_write(char *hash, size_t hash_len, bool confirm,
+				struct mcumgr_image_state *res_buf)
+{
+	struct net_buf *nb;
+	int rc;
+	uint32_t map_count;
+	zcbor_state_t zse[CBOR_ENCODER_STATE_NUM];
+	bool ok;
+
+	k_mutex_lock(&mcumgr_mutex, K_FOREVER);
+	image_info = res_buf;
+
+	/* allocate buffer */
+	nb = smp_client_buf_allocation(MGMT_GROUP_ID_IMAGE, IMG_MGMT_ID_STATE, MGMT_OP_WRITE);
+	if (!nb) {
+		rc = MGMT_ERR_ENOMEM;
+		goto end;
+	}
+
+	zcbor_new_encode_state(zse, ARRAY_SIZE(zse), nb->data + nb->len, net_buf_tailroom(nb), 0);
+	if (hash) {
+		map_count = 4;
+	} else {
+		map_count = 2;
+	}
+
+	/* Write map start init and confirm params */
+	ok = zcbor_map_start_encode(zse, map_count) && zcbor_tstr_put_lit(zse, "confirm") &&
+	     zcbor_bool_put(zse, confirm);
+	/* Write hash data */
+	if (ok && hash) {
+		ok = zcbor_tstr_put_lit(zse, "hash") && zcbor_bstr_encode_ptr(zse, hash, hash_len);
+	}
+	/* Close map */
+	if (ok) {
+		ok = zcbor_map_end_encode(zse, map_count);
+	}
+
+	if (!ok) {
+		smp_packet_free(nb);
+		rc = MGMT_ERR_ENOMEM;
+		goto end;
+	}
+
+	nb->len = zse->payload - nb->data;
+	k_sem_reset(&spm_mutex);
+	rc = smp_client_send_cmd(nb, image_state_res_fn, &spm_mutex,
+				 CONFIG_SMP_CMD_DEFAULT_LIFE_TIME);
+	if (rc) {
+		smp_packet_free(nb);
+		goto end;
+	}
+	k_sem_take(&spm_mutex, K_FOREVER);
+end:
+	k_mutex_unlock(&mcumgr_mutex);
+	return rc;
+}
+
+int img_mgmt_client_state_read(struct mcumgr_image_state *res_buf)
+{
+	struct net_buf *nb;
+	int rc;
+	zcbor_state_t zse[CBOR_ENCODER_STATE_NUM];
+	bool ok;
+
+	k_mutex_lock(&mcumgr_mutex, K_FOREVER);
+	image_info = res_buf;
+
+	/* allocate buffer */
```

Comment:
Comment can go. Comments should explain what the point of the code is or why a particular method was chosen, it should not just say what the function call is doing when that's obvious from the name of the function itself